File: 2017/day-3/direction.dart
enum Direction { right, up, left, down }

class Pair {
  final int x, y;
  Pair(this.x, this.y);
}

Pair getShiftPair(Direction direction) {
  switch (direction) {
    case Direction.right:
      return new Pair(1, 0);
    case Direction.up:
      return new Pair(0, -1);
    case Direction.left:
      return new Pair(-1, 0);
    case Direction.down:
      return new Pair(0, 1);
    default: throw new Exception('Unsupported direction $direction');
  }
}

String getString(Direction direction) {
  switch (direction) {
    case Direction.right:
      return '→';
    case Direction.up:
      return '↑';
    case Direction.left:
      return '←';
    case Direction.down:
      return '↓';
    default: throw new Exception('Unsupported direction $direction');
  }
}

File: 2015/day-3/src/Direction.java
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

enum Direction {
  UP('^'),
  DOWN('v'),
  LEFT('>'),
  RIGHT('<');

  private char character;

  Direction(char character) {
    this.character = character;
  }

  static Direction from(char character) {
    for (Direction direction : values()) {
      if (direction.character == character) {
        return direction;
      }
    }
    throw new IllegalArgumentException(character + " is not associated with a direction.");
  }

  static Direction getOpposite(Direction direction) {
    switch (direction) {
      case UP:
        return DOWN;
      case DOWN:
        return UP;
      case LEFT:
        return RIGHT;
      case RIGHT:
        return LEFT;
    }
    throw new IllegalArgumentException();
  }

  static List<Direction> getAdjacentDirections(Direction direction) {
    return Arrays.stream(values()).filter(otherDirection ->
        otherDirection != direction && otherDirection != getOpposite(direction)
    ).collect(Collectors.toList());
  }
}
